File: v4/perl/vector.pl


#!/opt/perl_5.18.2/bin/perl

BEGIN {
    push ( @INC,"./generated_files/");
    push ( @INC,"./perl/");
}
use Data::Dumper;
use STDLib;
use Scalar::Util qw(reftype);
my $p = "\n[PERL CODE]:\t";

#=================================================
sub Print {
    my $vi = $_[0];
    print $p," ==============================  print vector \n";
    for($i = 1; $i <=5 ; $i++)
    {
        print $p , "$i , value = " , $vi->get($i-1);
    } 
    print "\n";
}
#=================================================

print $p,"perl code started here";
print $p," ==============================";
print $p," ==============================  create STDLib::vectori \n";
my $vi1 = new STDLib::vectori(5);
print $p, " vi1 = ", $vi1 , "\n";
print $p, " vi1.size() = ", $vi1->size() , "\n";

print $p," ==============================  add values to vector \n";

for($i = 1; $i <=5 ; $i++)
{
    #vi1[$i] = $i+ 5;
    $vi1->set($i-1 , $i+5  );
}

Print($vi1);

my $vi2 = new STDLib::vectori($vi1);
Print($vi2);

print $p," ==============================   double average(std::vector<int> v); \n";

print $p, " average(vi1) = ", STDLib::average($vi1) , "\n";

print $p," ==============================   std::vector<double> half(const std::vector<double>& v) ; \n";
my $vd1 = new STDLib::vectord(5);
for($i = 1; $i <=5 ; $i++)
{
    #vi1[$i] = $i+ 5;
    $vd1->set($i-1 , $i+20  );
}

Print($vd1);

my $vd2 = STDLib::half($vd1);

#Print($vd2);
#Can't call method "get" on unblessed reference at perl/vector.pl line 18.
#make: *** [vec] Error 255



print $p," ==============================   void halve_in_place(std::vector<double>& v) ; \n";

my $vd3 = new STDLib::vectord(5);
for($i = 1; $i <=5 ; $i++)
{
    #vi1[$i] = $i+ 5;
    $vd3->set($i-1 , $i+10  );
}

Print($vd3);

STDLib::halve_in_place($vd3);

Print($vd3);

print $p," ==============================   passing PERL array's ; \n";
my @arr1 = (20,30,40,50);
print $p," arr1 = ", @arr1, "\n";
#print $p, " average(arr1) = ", STDLib::average(@arr1) , "\n";
#RuntimeError Usage: average(v); at perl/vector.pl line 86.
#make: *** [vec] Error 255

#my $vd10 = STDLib::half(@arr1);
#RuntimeError Usage: half(v); at perl/vector.pl line 88.
#make: *** [vec] Error 255


#STDLib::halve_in_place(@arr1);
#RuntimeError Usage: halve_in_place(v); at perl/vector.pl line 90.
#make: *** [vec] Error 255

print $p," ==============================   passing PERL's List  ; \n";
#print $p, " average(arr1) = ", STDLib::average( 10,20,30,40 ) , "\n";
#RuntimeError Usage: average(v); at perl/vector.pl line 100.
#make: *** [vec] Error 255


#print $p," ==============================   bool operator==(const std::vector<double>& v1 , const std::vector<double>& v2 ) ;  \n";
#my $vd10 = new STDLib::vectord(5);
#for($i = 1; $i <=5 ; $i++)
#{
    #vi1[$i] = $i+ 5;
#    $vd10->set($i-1 , $i+20  );
#}
#my $compRes = STDLib::operator==($vd1 , $vd10);
#print $p, " STDLib::operator==(vd1 , vd10); = ", $compRes , "\n";
# not working

print $p," ==============================\n";
print $p," perl code ended here\n\n";
print $p," ==============================\n";




exit;

File: v4/python/vector.py
#!/opt/python/3.11/bin/python
# -*- coding: utf-8 -*-

import sys
sys.path.append("./generated_files/");
sys.path.append("./perl/");

import copy
import STDLib;

def Print(v):
    #xi = v.begin();
    i =0;
    #print( [x for x in v]);

    while(i < v.size() ):
        print (f" index= {i} , value = {v[i]} ");
        i=i+1;
        

#==================================================================    
nl = '\n [PYTHON CODE] : ';

print ( nl," Testing Basic Python class testing String.i");

print ( nl, "******************************************** Print module  STDLib \n" );
print ( nl," STDLib =  ", STDLib);
print ( nl," type(STDLib)  =  ", type(STDLib) );
print ( nl," dir(STDLib)  =  ", dir(STDLib) );

print ( nl, "******************************************** v1 = STDLib.vectori(5) \n" );
v1 = STDLib.vectori(5);

print ( nl," v1 =  ", v1);
print ( nl," v1.size() =  ", v1.size());
print ( nl," type(v1)  =  ", type(v1) );

for i in range (0,5):
    v1[i] = i;


print ( nl, "******************************************** Print(v1); \n" );
Print(v1);

print ( nl, "******************************************** double average(std::vector<int> v); \n" );
print ( nl," STDLib.average( v1) =  ", STDLib.average( v1));


print ( nl, "******************************************** v1 = STDLib.vectord(5) \n" );
v2 = STDLib.vectord(5);

print ( nl," v2 =  ", v2);
print ( nl," v2.size() =  ", v2.size());
print ( nl," type(v2)  =  ", type(v2) );

for i in range (0,5):
    v2[i] = i + 10.0;

print ( nl, "******************************************** Print(v2); \n" );
Print(v2);

print ( nl, "******************************************** std::vector<double> half(const std::vector<double>& v) ;  \n" );
v3 = STDLib.half( v2);
#Print(v3);
#Traceback (most recent call last):
#  File "<>/FAST3_Python/v4/python/vector.py", line 64, in <module>
#    Print(v3);
#    ^^^^^^^^^
#  File "<>/FAST3_Python/v4/python/vector.py", line 16, in Print
#    while(i < v.size() ):
#              ^^^^^^
#AttributeError: 'tuple' object has no attribute 'size'

print( [x for x in v3]);
print ( nl, "******************************************** void halve_in_place(std::vector<double>& v) ; \n" );
STDLib.halve_in_place( v2);
Print(v2);

print ( nl, "******************************************** bool operator==(const std::vector<double>& v1 , const std::vector<double>& v2 ); \n" );
#print ( nl," STDLib.operator==(v2,v3) =  ", STDLib.operator==(v2,v3));
#ttributeError: module 'STDLib' has no attribute 'operator'

#swig/../cpp/vect.h:20: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.


print ( nl," (v2 == v3) =  ",(v2 == v3));
#//NOTE: Does not call to C++ operator==(), it does python == comparision

#print ( nl," (v2 === v3) =  ",(v2 === v3));
print ( nl, "******************************************** del v1  \n" );
del v1;

print (nl, "***********  END OF PROGRAM *********************************");